### DISCUSSION TITLE
Open source firmware

### DISCUSSION THREAD

**Discussion by maks** [bangcorrupt/hacktribe#117] [General]
@bangcorrupt was wondering how far off you think it is to get a minimal "blink a light" open source firmware build running? 
I guess you at least need to figure out the memory map for the linker to use to make a binary, no matter which Lang (C, Rust, Zig etc) the code is written in along with the register for a light or is there something sitting on a i2c or spi bus that's handling the buttons, knobs, leds etc?

And since you're able to binary patch the existing closed firmware, I'm assuming there is no code signing required for the bootloader?

And of course you've mentioned you need to get a debugger for the blackfin to start working out how the arm talks to it to actually be able to have an open source firmware make any sound...

I ask because I'm no where near an expert, but have done some embedded arm (cortex-m's) dev in the past, so would love to contribute if I can.

I wonder if as you work out the hardware details it would possible to create a qemu config for the arm to make it easier to do dev without needing to always run on a real e2 hardware?

**Comment by bangcorrupt** [comment]
I have a basic bootloader that will initialise the external RAM and read/write the flash, I'll share it once I work out what's broken.  I'd like to get it to the point where it will set up the hardware correctly for the factory firmware, there is quite a lot more to reverse here. 

All the panel IO comes via an ARM Cortex M3 over UART, I've started work reversing the protocol.  I have the debugger for the DSP, but haven't got as far as setting it up yet, may be a few weeks before I can do any real work on this.

**Reply by maks** [reply]
Wow nice work! I hadn't realised you had gotten that far already 👍
Yep that  makes sense, I've seen that alot using small Atmel or cortex-m doing IO into the "larger" Arm CPUs. I wonder why they used a M3 though, seems over powered for the job, unless maybe it does other IO jobs like SPI to the LCD or USB?

If it's just coming in over a serial connection, hopefully the protocol is fairly simple and I guess you can watch in a debugger attached and see what comes in as you press buttons, turn knobs etc.

This is all great news! and of course I'm not trying to rush you, appreciate we're all just doing this in our spare time.

In meantime I have some ideas about doing some things via sysex so will post here if I get it working.

**Reply by bangcorrupt** [reply]
~I think it does 2 UART (one for controls, the other for TRS MIDI) and some GPIO (for sync pulse and plug sensing).  I think~ the LCD is connected directly to the AM1802 CPU, the USB is definitely direct to the CPU. 

The UART needs the M3 initialised before it starts passing data to the CPU, I think I'm not far off getting this working.  The UART for panel IO is accessible using J3, shown bottom centre in [this photo](https://1.bp.blogspot.com/-4ZRj5FRodss/VcfOcHV2VoI/AAAAAAAAAr4/P0Ma_-C6KSA/s1600/IMG_9300.JPG).
